File: repertoire/concepts/data-integration/providers/enricher_plugin/OcrTesseract.swift
// Clef Data Integration Kit - Local OCR via Tesseract enricher provider
// Shells out to the tesseract binary, parses HOCR output for word-level bounding boxes.

import Foundation

public let OcrTesseractProviderID = "ocr_tesseract"
public let OcrTesseractPluginType = "enricher_plugin"

public struct ContentItem {
    public let id: String
    public let content: String
    public let contentType: String
    public let metadata: [String: Any]?
}

public struct EnricherConfig {
    public let model: String?
    public let apiKey: String?
    public let threshold: Double?
    public let options: [String: Any]?
}

public struct EnrichmentResult {
    public let fields: [String: Any]
    public let confidence: Double
    public let metadata: [String: Any]?
}

public struct SchemaRef {
    public let name: String
    public let fields: [String]?
}

public struct CostEstimate {
    public let tokens: Int?
    public let apiCalls: Int?
    public let durationMs: Int?
}

public struct WordBox {
    public let word: String
    public let x1: Int
    public let y1: Int
    public let x2: Int
    public let y2: Int
    public let confidence: Double
}

public enum EnricherError: Error {
    case processError(String)
    case parseError(String)
    case ioError(String)
}

public final class OcrTesseractEnricherProvider {
    public init() {}

    public func enrich(item: ContentItem, config: EnricherConfig) async throws -> EnrichmentResult {
        let language = (config.options?["language"] as? String) ?? "eng"
        let psm = (config.options?["psm"] as? Int) ?? 3
        let dpi = (config.options?["dpi"] as? Int) ?? 300

        // Decode base64 image and write to temp file
        guard let imageData = Data(base64Encoded: item.content) else {
            throw EnricherError.parseError("Failed to decode base64 image content")
        }

        let tmpPath = NSTemporaryDirectory() + "clef_ocr_\(item.id)_\(ProcessInfo.processInfo.processIdentifier).png"
        let tmpURL = URL(fileURLWithPath: tmpPath)

        try imageData.write(to: tmpURL)
        defer { try? FileManager.default.removeItem(at: tmpURL) }

        // Shell out to tesseract with HOCR output
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/usr/bin/env")
        process.arguments = [
            "tesseract", tmpPath, "stdout",
            "-l", language,
            "--psm", String(psm),
            "--dpi", String(dpi),
            "hocr"
        ]

        let stdoutPipe = Pipe()
        let stderrPipe = Pipe()
        process.standardOutput = stdoutPipe
        process.standardError = stderrPipe

        try process.run()
        process.waitUntilExit()

        let stdoutData = stdoutPipe.fileHandleForReading.readDataToEndOfFile()
        let stderrData = stderrPipe.fileHandleForReading.readDataToEndOfFile()

        guard process.terminationStatus == 0 else {
            let errMsg = String(data: stderrData, encoding: .utf8) ?? "Unknown error"
            throw EnricherError.processError("Tesseract failed: \(errMsg)")
        }

        let hocr = String(data: stdoutData, encoding: .utf8) ?? ""
        let (text, wordBoxes) = parseHocr(hocr)

        let avgConfidence: Double = wordBoxes.isEmpty ? 0.0 :
            wordBoxes.reduce(0.0) { $0 + $1.confidence } / Double(wordBoxes.count)

        let boxDicts: [[String: Any]] = wordBoxes.map { box in
            [
                "word": box.word,
                "x1": box.x1, "y1": box.y1,
                "x2": box.x2, "y2": box.y2,
                "confidence": box.confidence
            ]
        }

        return EnrichmentResult(
            fields: [
                "extracted_text": text,
                "word_boxes": boxDicts,
                "word_count": wordBoxes.count
            ],
            confidence: avgConfidence,
            metadata: [
                "provider": OcrTesseractProviderID,
                "language": language,
                "psm": psm,
                "dpi": dpi,
                "processingEngine": "tesseract"
            ]
        )
    }

    public func appliesTo(schema: SchemaRef) -> Bool {
        let imageSchemas = ["image", "document_scan", "scanned_page", "photo"]
        let nameLower = schema.name.lowercased()
        return imageSchemas.contains { nameLower.contains($0) }
    }

    public func costEstimate(item: ContentItem) -> CostEstimate {
        let sizeKb = Double(item.content.count) * 3.0 / 4.0 / 1024.0
        let estimatedMs = max(500.0, min(30000.0, sizeKb * 2.0))
        return CostEstimate(tokens: nil, apiCalls: 0, durationMs: Int(estimatedMs))
    }

    // Parse HOCR XML output for word-level bounding boxes
    private func parseHocr(_ hocr: String) -> (String, [WordBox]) {
        var wordBoxes: [WordBox] = []
        var textParts: [String] = []

        // Match ocrx_word spans with bbox and confidence
        let pattern = #"<span[^>]*class=['\"]ocrx_word['\"][^>]*title=['\"]bbox\s+(\d+)\s+(\d+)\s+(\d+)\s+(\d+);\s*x_wconf\s+(\d+)['\"][^>]*>([^<]+)</span>"#
        guard let regex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive]) else {
            return ("", [])
        }

        let nsHocr = hocr as NSString
        let matches = regex.matches(in: hocr, range: NSRange(location: 0, length: nsHocr.length))

        for match in matches {
            guard match.numberOfRanges >= 7 else { continue }
            let x1 = Int(nsHocr.substring(with: match.range(at: 1))) ?? 0
            let y1 = Int(nsHocr.substring(with: match.range(at: 2))) ?? 0
            let x2 = Int(nsHocr.substring(with: match.range(at: 3))) ?? 0
            let y2 = Int(nsHocr.substring(with: match.range(at: 4))) ?? 0
            let conf = Int(nsHocr.substring(with: match.range(at: 5))) ?? 0
            let word = nsHocr.substring(with: match.range(at: 6)).trimmingCharacters(in: .whitespaces)

            if !word.isEmpty {
                textParts.append(word)
                wordBoxes.append(WordBox(
                    word: word,
                    x1: x1, y1: y1, x2: x2, y2: y2,
                    confidence: Double(conf) / 100.0
                ))
            }
        }

        return (textParts.joined(separator: " "), wordBoxes)
    }
}

File: repertoire/concepts/data-integration/providers/enricher_plugin/ocr_tesseract.ts
// Clef Data Integration Kit - Local OCR via Tesseract enricher provider
// Shells out to the tesseract binary, parses HOCR output for word-level bounding boxes.

import { execFile } from 'child_process';
import { promisify } from 'util';
import { writeFileSync, unlinkSync } from 'fs';
import { tmpdir } from 'os';
import { join } from 'path';

const execFileAsync = promisify(execFile);

export const PROVIDER_ID = 'ocr_tesseract';
export const PLUGIN_TYPE = 'enricher_plugin';

export interface ContentItem {
  id: string;
  content: string;
  contentType: string;
  metadata?: Record<string, unknown>;
}

export interface EnricherConfig {
  model?: string;
  apiKey?: string;
  threshold?: number;
  options?: Record<string, unknown>;
}

export interface EnrichmentResult {
  fields: Record<string, unknown>;
  confidence: number;
  metadata?: Record<string, unknown>;
}

export interface SchemaRef {
  name: string;
  fields?: string[];
}

export interface CostEstimate {
  tokens?: number;
  apiCalls?: number;
  durationMs?: number;
}

interface WordBox {
  word: string;
  x1: number;
  y1: number;
  x2: number;
  y2: number;
  confidence: number;
}

function parseHocr(hocr: string): { text: string; wordBoxes: WordBox[] } {
  const wordBoxes: WordBox[] = [];
  const lines: string[] = [];

  // Match ocrx_word spans: <span class='ocrx_word' ... title='bbox x1 y1 x2 y2; x_wconf NN'>word</span>
  const wordPattern = /<span[^>]*class=['"]ocrx_word['"][^>]*title=['"]bbox\s+(\d+)\s+(\d+)\s+(\d+)\s+(\d+);\s*x_wconf\s+(\d+)['"][^>]*>([^<]+)<\/span>/gi;
  let match: RegExpExecArray | null;

  while ((match = wordPattern.exec(hocr)) !== null) {
    const [, x1Str, y1Str, x2Str, y2Str, confStr, word] = match;
    const conf = parseInt(confStr, 10) / 100;
    wordBoxes.push({
      word: word.trim(),
      x1: parseInt(x1Str, 10),
      y1: parseInt(y1Str, 10),
      x2: parseInt(x2Str, 10),
      y2: parseInt(y2Str, 10),
      confidence: conf,
    });
  }

  // Extract line-level text from ocr_line spans
  const linePattern = /<span[^>]*class=['"]ocr_line['"][^>]*>([\s\S]*?)<\/span>\s*(?=<span[^>]*class=['"]ocr_line|<\/p>|<\/div>)/gi;
  let lineMatch: RegExpExecArray | null;
  while ((lineMatch = linePattern.exec(hocr)) !== null) {
    const lineHtml = lineMatch[1];
    const stripped = lineHtml.replace(/<[^>]+>/g, ' ').replace(/\s+/g, ' ').trim();
    if (stripped) lines.push(stripped);
  }

  const text = lines.length > 0 ? lines.join('\n') : wordBoxes.map((w) => w.word).join(' ');
  return { text, wordBoxes };
}

export class OcrTesseractEnricherProvider {
  async enrich(item: ContentItem, config: EnricherConfig): Promise<EnrichmentResult> {
    const language = (config.options?.language as string) ?? 'eng';
    const psm = (config.options?.psm as number) ?? 3;
    const dpi = (config.options?.dpi as number) ?? 300;

    // Write image content (base64-encoded) to a temp file
    const tmpPath = join(tmpdir(), `clef_ocr_${item.id}_${Date.now()}.png`);
    const imageBuffer = Buffer.from(item.content, 'base64');
    writeFileSync(tmpPath, imageBuffer);

    try {
      const args = [
        tmpPath,
        'stdout',
        '-l', language,
        '--psm', String(psm),
        '--dpi', String(dpi),
        'hocr',
      ];

      const { stdout, stderr } = await execFileAsync('tesseract', args, {
        timeout: 60000,
        maxBuffer: 10 * 1024 * 1024,
      });

      if (stderr && stderr.includes('Error')) {
        throw new Error(`Tesseract error: ${stderr}`);
      }

      const { text, wordBoxes } = parseHocr(stdout);
      const avgConfidence = wordBoxes.length > 0
        ? wordBoxes.reduce((sum, w) => sum + w.confidence, 0) / wordBoxes.length
        : 0;

      return {
        fields: {
          extracted_text: text,
          word_boxes: wordBoxes,
          word_count: wordBoxes.length,
        },
        confidence: avgConfidence,
        metadata: {
          provider: PROVIDER_ID,
          language,
          psm,
          dpi,
          processingEngine: 'tesseract',
        },
      };
    } finally {
      try { unlinkSync(tmpPath); } catch { /* cleanup best-effort */ }
    }
  }

  appliesTo(schema: SchemaRef): boolean {
    const imageSchemas = ['image', 'document_scan', 'scanned_page', 'photo'];
    return imageSchemas.some((s) => schema.name.toLowerCase().includes(s));
  }

  costEstimate(item: ContentItem): CostEstimate {
    const sizeKb = Buffer.byteLength(item.content, 'base64') / 1024;
    // Tesseract local processing: ~500ms per page, scales with image size
    const estimatedMs = Math.max(500, Math.min(30000, sizeKb * 2));
    return {
      durationMs: Math.round(estimatedMs),
      apiCalls: 0,
    };
  }
}

export default OcrTesseractEnricherProvider;
